Replaces plurality voting with approval voting for some votes

Putnam3145 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

As the title: replaces the [first past the post](https://en.wikipedia.org/wiki/First-past-the-post_voting) voting method used in the codebase right now with [approval voting](https://en.wikipedia.org/wiki/Approval_voting).

## Why It's Good For The Game

First past the post sucks. I mean, it's just awful. Sometimes the winner is someone who the actual majority of voters doesn't want, it's highly susceptible to strategic voting, it leads to a massive amount of similar perverse results...

Approval voting is simple: you can vote for as many options as you want. For two-choice votes, it's identical, as long as there's one winner; voting both options is the same as voting one option. For multi-winner votes, it's *still* better, because voting both makes the vote closer, which is what you'd *want*.

And if you want to know why approval over IRV: well, [see this article, if you have the time](https://www.electionscience.org/library/approval-voting-versus-irv/). That, plus it's much, much easier to implement approval (see how small this change is!)

## Changelog
:cl:
tweak: Voting now uses approval instead of first-past-the-post
/:cl:

Putnam3145 committed:
Replaces FPTP voting with approval voting for all votes

nemvar commented:
config would make more sense imo

Putnam3145 commented:
I'll do it if that's the consensus obv. (by which I mean one more person tells me to do it) but I will argue it's not really necessary to do so, since the votes are basically counted the same way and... I really can't emphasize enough how garbage first-past-the-post is.

nemvar commented:
This system is used for more than just map votes.
It could be used to run a quick "Yes, No, Maybe" poll in which plurality makes no sense.

Putnam3145 commented:
I know it's used for more than just map votes, I pointed out that it's still basically identical for yes/no votes, too.

Bokkiewokkie commented:
> This system is used for more than just map votes.
> It could be used to run a quick "Yes, No, Maybe" poll in which plurality makes no sense.

This is why I think it should, if it gets added, be an optional thing to add to a poll and not a permanent change to all polls forever.

cacogen commented:
Bokkiewokkie's right, on custom votes admins should be able to specify how many choices can be selected. Map votes and gamemode votes should allow as many choices as maps, and restart votes should only allow one choice.

Putnam3145 commented:
Alright, yeah, that's all fine. Conveniently, I already have all that code in place from last time I implemented this, but I decided to keep it simple for this PR because that was hellcode (I also implemented: IRV, Schulze, score, majority judgement, and allowed any of them for custom polls. majority judgement and IRV used to work, but don't anymore. I don't know what happened. I don't want to repeat that.)

Putnam3145 committed:
Made vote type vote-to-vote basis

Putnam3145 commented:
Alright, suggestions implemented.

spookydonut commented on code/__DEFINES/vote.dm:
```diff
@@ -0,0 +1,7 @@
+#define PLURALITY_VOTING 0
+#define APPROVAL_VOTING 1
+
+GLOBAL_LIST_INIT(vote_type_names,list(\
```
why is this a global list

Putnam3145 replied:
Shit, I mindlessly copied that from my Cit implementation where (IIRC) I used it more than once. Yeah, I'll just make it in the actual proc.

Putnam3145 committed:
Moved vote_type_names from being a glob

Putnam3145 commented:
Yeah, if y'all got anything else I'd be happy to implement it.

cacogen commented:
pls merge

PKPenguin321 commented:
This is for votes and not polls, yes?

Putnam3145 commented:
Yes, exclusively for in-game votes, started by the "vote" verb or by the server.

cacogen commented:
still merge. makes map votes less skewed towards meta

Rohesie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -157,22 +158,43 @@ SUBSYSTEM_DEF(vote)
+						voted[usr.ckey] = list(vote)
+						choices[choices[vote]]++
+						return vote
+				else // plurality-by-default
```
`if(PLURALITY_VOTING )`
and
`else CRASH()`

Rohesie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -157,22 +158,43 @@ SUBSYSTEM_DEF(vote)
+						choices[choices[vote]]++
+						return vote
+				else // plurality-by-default
+					if(usr.ckey in voted)
```
```suggestion
					if(voted[usr.ckey])
```
Faster lookup.

Rohesie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -157,22 +158,43 @@ SUBSYSTEM_DEF(vote)
+		if(vote && ISINRANGE(vote, 1, choices.len))
+			switch(vote_system)
+				if(APPROVAL_VOTING)
+					if(usr.ckey in voted)
```
```suggestion
					if(voted[usr.ckey])
```

Rohesie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -203,7 +225,10 @@ SUBSYSTEM_DEF(vote)
 				question = stripped_input(usr,"What is the vote for?")
 				if(!question)
-					return FALSE
+					return 0
```
Why are you returning zero instead of `FALSE`?

Rohesie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -203,7 +225,10 @@ SUBSYSTEM_DEF(vote)
 				if(!question)
-					return FALSE
+					return 0
+				var/list/vote_type_names = list("Plurality (can only vote one option)" = PLURALITY_VOTING,"Approval (can vote any amount)" = APPROVAL_VOTING)
+				var/system_string = input(usr,"Which voting type?",vote_type_names[1]) in vote_type_names
```
Aren't you giving here a single choice?

ExcessiveUseOfCobblestone commented:
20 days to fix code, most of it suggestion blocks, would suggest this is not being updated.

cacogen commented:
rip
